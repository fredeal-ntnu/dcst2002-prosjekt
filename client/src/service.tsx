import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type Question = {
  question_id: number;
  title: string;
  text: string;
  view_count: number;
  has_answer: boolean;
  user_id: string;
};

export type Tag = {
  tag_id: number;
  name: string;
};

export type Tag_Question_Relation = {
  tag_id: number;
  question_id: number;
};

export type User = {
  user_name: string;
  password: string;
};

export type Answer = {
  answer_id: number;
  text: string;
  confirmed_answer: boolean;
  question_id: number;
  score?: number;
};

export type QuestionComment = {
  question_comment_id: number;
  text: string;
  question_id: number;
};

export type AnswerComment = {
  answer_comment_id: number;
  text: string;
  answer_id: number;
};

export type Vote = {
  user_id: number;
  answer_id: number;
  vote_type: boolean;
}

class Service {
  /**
   * Get all questions.
   */

  getAllQuestions() {
    return axios.get<Question[]>('/questions').then((response) => response.data);
  }

  /**
   * Get a question by id.
   */

  getQuestion(id: number) {
    return axios.get<Question>('/questions/' + id).then((response) => response.data);
  }

  /**
   * Get top five questions.
   */

  getTopFiveQuestions() {
    return axios.get<Question[]>('/topfivequestions').then((response) => response.data);
  }

  /**
   * Get all unaswered questions.
   */
  getUnansweredQuestions() {
    return axios.get<Question[]>('/unansweredquestions').then((response) => response.data);
  }

  /**
   * Update a question.
   */

  updateQuestion(question: Question) {
    return axios.put('/questions', question).then((response) => response.data);
  }

  /**
   * Create a question.
   */

  createQuestion(title: string, text: string, user_id: number) {
    return axios.post('/questions', { title, text, user_id }).then((response) => response.data.id);
  }

  /**
   * Delete a question.
   */

  deleteQuestion(id: number) {
    console.log(typeof id);
    return axios.delete('/questions/' + id).then((response) => response.data);
  }

  /**
   * Get tags for a question.
   */

  getTags(id: number) {
    return axios.get('/questions/' + id + '/tags').then((response) => response.data);
  }

  /**
   * Get all tags.
   */

  getAllTags() {
    return axios.get('/tags').then((response) => response.data);
  }

  createTagQuestionRelation(tag_id: number, question_id: number) {
    return axios.post('/questions/' + question_id, { tag_id, question_id });
  }

  /**
   * Get a tag by name.
   */

  getQuestionsByTagId(tag: number) {
    return axios.get('/tag/' + tag + '/questions').then((response) => response.data);
  }

  //get all tags for a question by question id
  getTagsForQuestion(id: number) {
    return axios.get('/questions/' + id + '/tags').then((response) => response.data);
  }

  getAllTagQuestionsRelations() {
    return axios.get('/question/:id').then((response) => response.data);
  }

  //get a user
  getUser(user: User) {
    return axios.get<User>('/users/' + user.user_name).then((response) => response.data);
  }

  getAnswerById(id: number) {
    return axios.get('/answers/' + id).then((response) => response.data);
  }

  getAnswersForQuestion(id: number) {
    return axios.get('/questions/' + id + '/answers').then((response) => response.data);
  }

  createAnswer(text: string, question_id: number) {
    return axios
      .post('/questions/' + question_id + '/answers', { text, question_id })
      .then((response) => response.data);
  }

  // ALLE SERVICES FOR QUESTION COMMENTS

  getQuestionCommentById(id: number) {
    return axios
    .get('/comments/'+ id)
    .then((response) => response.data);
  }

  createQuestionComment(text: string, question_id: number) {
    return axios
      .post('/questions/' + question_id + '/comments', { text, question_id })
      .then((response) => response.data);
  }

updateQuestionComment(questionComment: QuestionComment) {
  return axios
  .put('/comments/', questionComment)
  .then((response) => response.data);
}

  deleteQuestionComment(id: number) {
    return axios
      .delete('/comments/' + id)
      .then((response) => response.data);
  }

  getQuestionCommentsForQuestion(id: number) {
    return axios
    .get('/questions/' + id + '/comments')
    .then((response) => response.data);
  }

  // ALLE SERVICES FOR ANSWER COMMENTS
  

  getAnswerCommentsForAnswer(id: number) {
    return axios.get('/answers/' + id + '/comments')
    .then((response) => response.data);
  }

  createAnswerComment(text: string, answer_id: number) {
    return axios
      .post('/answers/' + answer_id + '/comments', { text, answer_id })
      .then((response) => response.data);
  }

  getAnswerCommentById(id: number) {
    console.log('andre')
    return axios
    .get('/answer/comments/' + id)
    .then((response) => response.data);
  }

  updateAnswerComment(answerComment: AnswerComment) {
    return axios
    .put('/comments/', answerComment)
    .then((response) => response.data);
  }

  deleteAnswerComment(id: number) {
    return axios
      .delete('/answer/comments/' + id)
      .then((response) => response.data);
  }

  


  
  updateAnswer(answer: Answer) {
    return axios.put('/answers', answer)
    .then((response) => response.data);
  }

deleteAnswer(id: number) {
  return axios
  .delete('/answers/' + id)
  .then((response) => response.data);
}



// alle services for votes

getVotesByAnswerId(id: number) {
  return axios
  .get('/answers/' + id + '/votes')
  .then((response) => response.data);

}

createVoteForAnswer(user_id: number, answer_id: string, vote_type: boolean) {
  return axios
  .post(answer_id + '/votes', { user_id, answer_id, vote_type })
  .then((response) => response.data);
}


}

const service = new Service();
export default service;
