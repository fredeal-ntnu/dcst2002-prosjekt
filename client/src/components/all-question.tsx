import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Alert, Column, Row, Form } from '../widgets';
import service, { Question, Tag } from '../service';

export class AllQuestions extends Component {
    questions: Question[] = [];
    filter: string = 'All'; // State to manage the filter type
    search = '';
    tags: Tag[] = [];
  
    render() {
      return (
        <>   
          <Card title='All Questions'>
            <Row>
                <Column width={2}>
                <Form.Select value={this.filter} onChange={this.handleFilterChange}>
                    <option value="all">All Questions</option>
                    <option value="popular">Most Popular</option>
                    <option value="unanswered">Unanswered</option>
                </Form.Select>
                </Column>
                <Column width={10}>
                    <Form.Input
                    type='text' 
                    value={this.search} 
                    onChange={(event) => this.search = event.currentTarget.value}
                    />
                </Column>
            </Row>

               
            {/* <Row>
            {this.tags.map((tag) => (
                        <Column>
                            <Form.Label>
                                {tag.name}
                                <input type='checkbox'
                                    value={tag.tag_id}
                                    onChange={() => {'vet ikke'}}
                                />
                            
                            </Form.Label>
                        
                        </Column>
                    ))}
            </Row> */}
          </Card>
          {this.questions
                .filter((question) => (question.title.toLowerCase().includes(this.search.toLowerCase())))
                .map((question, i) => (
                    <Card key={i} title={question.title}>
                        <Row>
                            <Column width={2}>{question.text}</Column>
                        </Row>
                    </Card>
                ))
                }
        </>
      );
    }

    handleFilterChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
        this.filter = event.target.value;
        this.loadQuestions(); // Call a method to load questions based on the selected filter
      };
  
    // Method to load questions based on the current filter
    loadQuestions() {
      // Add logic here to fetch and filter questions accordingly
      // This is a placeholder for whatever your service methods might be
      switch (this.filter) {
        case 'All':
          // Fetch all questions
          break;
        case 'Popular':
          // Fetch most popular questions
          break;
        case 'Unanswered':
          // Fetch unanswered questions
          break;
        default:
          // Handle default case
      }
    }
  
    // Lifecycle method to load questions when the component mounts
    mounted() {
      this.loadQuestions();

      service
        .getAllTags()
        .then((tags) => (this.tags = tags))
        .catch((error) => Alert.danger(error.message));  
      
      service
        .getAllQuestions()
        .then((questions) => (this.questions = questions))
        .catch((error) => Alert.danger(error.message));

  }
}