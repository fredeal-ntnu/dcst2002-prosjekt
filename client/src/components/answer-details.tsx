import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Row, Column, Button, Form } from '../widgets';
import service, { Answer, AnswerComment, User } from '../service';
import { createHashHistory } from 'history';

const history = createHashHistory();

export class AnswerDetails extends Component<{ match: { params: { id: number } } }> {
  answerComments: AnswerComment[] = [];
  answer: Answer = {
    answer_id: 0,
    text: '',
    confirmed_answer: 0,
    last_updated: new Date(),
    question_id: 0,
    user_id: 0,
  };
  answerComment: AnswerComment = { answer_comment_id: 0, text: '', answer_id: 0, user_id: 0 };
  user: User = { user_id: 0, google_id: '', username: '', email: '' };
  connectedUser: number = 0;

  render() {
    return (
      <>
        <Card title="Answer">
          <Row>
            <Column width={10}>{this.answer.text}</Column>
          </Row>
        </Card>
        <Card title="Comments">
          {this.answerComments.map((answerComment) => (
            <Row key={answerComment.answer_comment_id}>
              <Column width={10}>
                {answerComment.text}
                <Column>
                  <Button.Success
                    onClick={() =>
                      history.push(
                        '/questions/' +
                          this.answer.question_id +
                          '/answers/' +
                          answerComment.answer_id +
                          '/comments/' +
                          answerComment.answer_comment_id +
                          '/edit',
                      )
                    }
                  >
                    Edit
                  </Button.Success>
                </Column>
              </Column>
            </Row>
          ))}
        </Card>

        <Card title="Add Comment">
          <Form.Textarea
            placeholder="Add comment"
            type="text"
            value={this.answerComment.text}
            onChange={(event) => (this.answerComment.text = event.currentTarget.value)}
            maxLength={1000}
            style={{ width: '500px' }}
          />
          <Column>
            <Button.Success onClick={this.addAnswerComment}>Add Comment</Button.Success>
          </Column>
        </Card>
      </>
    );
  }

  mounted() {
    service

      //Get logged in user
      .getMe()
      .then((user) => {
        this.user = user;
        this.connectedUser = this.user.user_id;
      })
      .catch((error) => {
        console.error(error.message);
        history.push('/');
        alert('You must be logged in to access this page');
      });

    service
      .getAnswerById(this.props.match.params.id)
      .then((answer) => {
        this.answer = answer;
        return service.getAnswerCommentsForAnswer(this.answer.answer_id);
      })
      .then((answerComments) => (this.answerComments = answerComments))
      .catch((error: Error) => console.error('Error getting answer: ' + error.message));
  }

  addAnswerComment() {
    
    if(this.connectedUser) {
      service
      .createAnswerComment(this.answerComment.text, this.answer.answer_id, this.connectedUser)
      .then(() => this.mounted())
      .then(() => (this.answerComment.text = ''))
      .catch((error) => console.error('Error adding answer comment: ' + error.message));
    }else return
    
  }
}
