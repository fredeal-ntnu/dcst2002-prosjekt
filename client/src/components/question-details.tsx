import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Alert, Row, Column, Button, SideMenu, Form } from '../widgets';
import service, {
  Question,
  Tag,
  Tag_Question_Relation,
  Answer,
  QuestionComment,
  AnswerComment,
  Vote,
  User,
} from '../service';
import { createHashHistory } from 'history';

const history = createHashHistory();

interface State {
  isFavorite: boolean;
  isConfirmedAnswer: boolean;
  // define other state properties here
}

export class QuestionDetails extends Component<{ match: { params: { id: number } } }> {
  relations: Tag_Question_Relation[] = [];
  tags: Tag[] = [];
  answers: Answer[] = [];
  questionComments: QuestionComment[] = [];
  filter: string = 'all'; // State to manage the filter type

  votes: Vote[] = [];
  question: Question = {
    question_id: 0,
    title: '',
    text: '',
    view_count: 0,
    has_answer: 0,
    user_id: 0,
  };
  answer: Answer = {
    answer_id: 0,
    text: '',
    confirmed_answer: 0,
    last_updated: new Date(),
    question_id: 0,
    user_id: 0,
  };
  questionComment: QuestionComment = { question_comment_id: 0, text: '', question_id: 0, user_id: 0};
  answerComment: AnswerComment = { answer_comment_id: 0, text: '', answer_id: 0, user_id: 0};
  vote: Vote = { user_id: 0, answer_id: 0, vote_type: 0 };
  user: User = { user_id: 0, google_id: '', username: '', email: '' };
  score: number = 0;
  connectedUser: number = 0;
  answers_votes: Answer[] = [];

  state: State = {
    isFavorite: false,
    isConfirmedAnswer: false,
  };

  render() {
    return (
      <>
      <Column width={3}>
        <SideMenu header='Menu'
            items={[
              { label: "Questions", to: "/questions" },
              { label: "My Questions", to: "/myquestions" },
              { label: "My Favourite Answers", to: "/favourites" },
              { label: "New Question", to: "/createquestion" },
              { label: "Tags", to: "/tags" }
            ]}/>
        </Column>
        <Card title="Question">
          <Card title="Title">{this.question.title}</Card>

          <Card title="Text">{this.question.text}</Card>

          <Row>
            <Card title="Tags">
              {this.relations
                .filter((relation) => relation.question_id == this.props.match.params.id)
                .map((relation) => this.tags.find((tag) => tag.tag_id == relation.tag_id))
                .filter((tag): tag is Tag => tag !== undefined)
                .map((tag) => tag.name)
                .join(', ')}
            </Card>
          </Row>
          <Row>
            <Column width={8}>
             {this.createQuestionEditButton()}
            </Column>
          </Row>
        </Card>
        <Row>
          <Column width={6}>
          {this.handleQuestionCommentDisplay()}
          {this.addQuestionCommentInput()}
          </Column>
          <Column width={6}>
            {this.handleAnswerMapDisplay()}
            {this.addAnswerInput()}
          </Column>  
        </Row>
        
      </>
    );
  }

  mounted() {
    service.getMe().then((user) => {
      this.user = user;
      this.connectedUser = user.user_id;
    });
    service
      .getAllTags()
      .then((tags) => (this.tags = tags))
      .catch((error) => console.error(error.message));

    service
      .getQuestion(this.props.match.params.id)
      .then((question) => (this.question = question))
      .catch((error: Error) => console.error('Error getting question: ' + error.message));

    service.getAllTagQuestionRelations()
    .then((relations) => (this.relations = relations));

    service
      .getAnswerScoresByQuestionId(this.props.match.params.id)
      .then((answers_votes) => (this.answers_votes = answers_votes))
      .catch((error: Error) => console.error('Error getting answers: ' + error.message));

    service
      .getQuestionCommentsForQuestion(this.props.match.params.id)
      .then((questionComments) => (this.questionComments = questionComments))
      .catch((error: Error) => console.error('Error getting question comments: ' + error.message));


      //DENNE FUNKER MED Ã… HENTE VOTES FOR EN ANSWER
    // service
    // .getAnswersByQuestionId(this.props.match.params.id)
    // .then((answers) => (this.answers = answers))
    // .catch((error: Error) => console.error('Error getting answers: ' + error.message))

    // this.answers.map((answer) => {
    //   service.getVotesByAnswerId(answer.answer_id)
    //     .then((vote) => { this.vote = vote })
    //     .catch((error) => {
    //       console.error(error);
    //     });
    // });
  }


  handleFilterChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    console.log(event.target.value);
    this.filter = event.target.value;
    this.loadAnswers() // Call a method to load questions based on the selected filter
  };

 
  loadAnswers() {
    console.log(this.answers)
    // Add logic here to fetch and filter questions accordingly
    // This is a placeholder for whatever your service methods might be
    switch (this.filter) {
      case 'all':
        service
        .getAnswerScoresByQuestionId(this.props.match.params.id)
        .then((answers_votes) => (this.answers_votes = answers_votes))
        .catch((error: Error) => console.error('Error getting answers: ' + error.message));
        break;
      case 'popular':
        service
          .getAnswerScoresByQuestionId(this.props.match.params.id)
          .then((answers_votes) => {
            // Sort answers by score in descending order
            this.answers_votes = answers_votes.sort((a: any, b: any) => b.score - a.score);
          })
          .catch((error: Error) => console.error('Error getting answers: ' + error.message));
        break;
      case 'mostRecent':
        service
          .getAnswerScoresByQuestionId(this.props.match.params.id)
          .then((answers_votes) => {
            // Sort answers by last_updated in descending order
            this.answers_votes = answers_votes.sort((a: any, b: any) => {
              const dateA = new Date(a.last_updated);
              const dateB = new Date(b.last_updated);
              return dateB.getTime() - dateA.getTime();
            });
          })
          .catch((error: Error) => console.error('Error getting answers: ' + error.message));
        break;
    }
  }
  

createQuestionEditButton() {
  
  if(this.connectedUser == this.question.user_id) {
    return(
      <Button.Success
      onClick={() =>  history.push('/questions/' + this.props.match.params.id + '/edit')}
    >
      Edit
    </Button.Success>
    )
  }
    else return
}


  addQuestionCommentInput() {
    if(this.connectedUser) {
      return(
        <Card title="Add comment">
        <Row>
          <Column>
            <Form.Textarea
              placeholder="Add comment"
              type="text"
              value={this.questionComment.text}
              onChange={(event) => (this.questionComment.text = event.currentTarget.value)}
              rows={5}
            />
          </Column>
        </Row>
        <Row>
          <Column>
            <Button.Success onClick={this.createComment}>Add</Button.Success>
          </Column>
        </Row>
      </Card>
      )
    }else return
  }

  
  createComment() {
    if(this.connectedUser) {
      service
      .createQuestionComment(this.questionComment.text, this.props.match.params.id, this.connectedUser)
      .then(() => this.mounted())
      .catch((error) => console.error('Error saving comment: ' + error.message));
    }
    else alert("You have to be logged in to comment")
  }
  addAnswerInput() {
    if(this.connectedUser) {
      return(
        <Card title="Add answer">
          <Row>
            <Column>
              <Form.Textarea
                placeholder="add answer"
                type="text"
                value={this.answer.text}
                onChange={(event) => (this.answer.text = event.currentTarget.value)}
                rows={5}
              />
            </Column>
          </Row>
          <Row>
            <Column>
              <Button.Success onClick={this.createAnswer}>Add</Button.Success>
            </Column>
          </Row>
          </Card>
      )
    }else return
  } 



handleQuestionCommentDisplay() {
  
  if(this.connectedUser) {
   return(
    <Card title="Comments">
    {this.questionComments.map((questionComment) => {
      if (questionComment.question_id == this.props.match.params.id) {
        return (
          <Card title="" key={questionComment.question_comment_id}>
            <Row>
              {questionComment.text}
              <Row>
                <Column>
                 {this.handleQuestionCommentEdit(questionComment)}
                </Column>
              </Row>
            </Row>
          </Card>
        );
      }
    })}
  </Card>
   )
  }
  else {
    return(
      <Card title="Comments">
      {this.questionComments.map((questionComment) => {
        if (questionComment.question_id == this.props.match.params.id) {
          return (
            <Card title="" key={questionComment.question_comment_id}>
              <Row>
                {questionComment.text}
              </Row>
            </Card>
          );
        }
      })}
    </Card>
    )
  }
}

handleQuestionCommentEdit(questionComment: QuestionComment) {
  if(this.connectedUser == questionComment.user_id) {
    return(
      <Button.Success
      onClick={() =>
        history.push(
          '/questions/' +
            this.props.match.params.id +
            '/comments/' +
            questionComment.question_comment_id +
            '/edit',
        )
      }
    >
      Edit
    </Button.Success>
    )
  } else return
}


handleAnswerMapDisplay () {
  if(this.connectedUser) {
return(
  <Card title="Answers HUSK SORTERING">
    <Form.Select value={this.filter} onChange={this.handleFilterChange}>
                    <option value="all">All Answers</option>
                    <option value="popular">Most Popular Answers</option>
                    <option value="mostRecent">Most Recent</option>
    </Form.Select>
    {this.answers_votes.map((answer) => {
      
      const isFavoriteKey = `isFavorite_${answer.answer_id}`;
      if (answer.question_id == this.props.match.params.id) {
        return (
          <Card title="" key={answer.answer_id}>
            <Row>
              {answer.text}
              <Row>
                <Column>
                  <Button.Success
                    onClick={() => {
                      this.addUpvote(answer.answer_id);
                    }}
                  >
                    Upvote
                  </Button.Success>
                </Column>
                <Column>
                  <Button.Success
                    onClick={() => {
                      this.addDownvote(answer.answer_id);
                    }}
                  >
                    Downvote
                  </Button.Success>
                </Column>
                <Column>
                  <Card title="Votes">{answer.score}</Card>
                </Column>
                <Column>
                  <Button.Success
                    onClick={() => {
                      this.sendToAnswerCommentPage(answer.answer_id);
                    }}
                  >
                    Comments
                  </Button.Success>
                </Column>
                <Column>
                {this.handleEditAnswer(answer.answer_id,answer.user_id)}
                </Column>
                <Column>
                  <Button.Success onClick={() => this.setConfirmedAnswer(answer.answer_id)}>
                    Mark as best
                  </Button.Success>
                </Column>
                <Column>
                  <Button.Success
                    onClick={() => {
                      if (this.state[isFavoriteKey as keyof State]) {
                        this.addFavourite(answer.answer_id, this.connectedUser);
                        this.setState({ [isFavoriteKey]: false });
                      } else {
                        this.addFavourite(answer.answer_id, this.connectedUser);
                        this.setState({ [isFavoriteKey]: true });
                      }
                    }}
                  >
                    {this.state[isFavoriteKey as keyof State] ? 'Remove from favorites' : 'Add to favorites'}
                  </Button.Success>
                </Column>
              </Row>
            </Row>
          </Card>
        );
      }
    })}
    
   
  </Card>
)
    

  }

  else return (
    <Card title="Answers">
    {this.answers_votes.map((answer) => {
      const isFavoriteKey = `isFavorite_${answer.answer_id}`;
      if (answer.question_id == this.props.match.params.id) {
        return (
          <Card title="" key={answer.answer_id}>
            <Row>
              {answer.text}
              
                
                <Column>
                  <Card title="Votes">{answer.score}</Card>
                </Column>
                
               
              </Row>
            
          </Card>
        );
      }
    })}
    
   
  </Card>
  )
}


handleEditAnswer(answer_id: number, user_id: number) {
  

  if(this.connectedUser == user_id) {
    return(
      <Button.Success
      onClick={() =>  history.push('/questions/' + this.props.match.params.id + '/answers/' + answer_id + '/edit')}
    >
      Edit
    </Button.Success>
    )
  }

  else return
}
  createAnswer() {
    if(this.connectedUser) {
      service
      .createAnswer(this.answer.text, this.props.match.params.id, this.connectedUser)
      .then(() => this.setHasAnswered())
      .then(() => this.mounted())
      .catch((error) => console.error('Error saving answer: ' + error.message));
    }
    else(alert("You have to be logged in to answer"))
  }
  async setHasAnswered() {
    this.question.has_answer = 1;
    await service
      .updateQuestion(this.question)
      .catch((error) => console.error('Error saving question: ' + error.message));
  }

  sendToAnswerCommentPage(answer_id: number) {
    history.push('/questions/' + this.props.match.params.id + '/answers/' + answer_id);
  }


  addUpvote(answer_id: number) {
    if(this.connectedUser) {
      service
      .createVote(this.connectedUser, answer_id, 1)
      .then(() => this.mounted())
      .catch((error) => console.error('Error saving answer: ' + error.message));
    }
    else(alert("You have to be logged in to vote"))
  }

  addDownvote(answer_id: number) {
    if(this.connectedUser) {
      service
      .createVote(this.connectedUser, answer_id, 0)
      .then(() => this.mounted())
      .catch((error) => console.error('Error saving answer: ' + error.message));
    }
    else alert("You have to be logged in to vote")
   
  }

  addFavourite(answer_id: number, user_id: number) {
    if(this.connectedUser) {
      service
      .handleFavouriteRelation(answer_id, user_id)
      .catch((error) => console.error('Error saving answer: ' + error.message));
    }
    else alert("You have to be logged in to add to favorites")
    
  }

  // Set confirmed answer for connected user

   setConfirmedAnswer(answer_id: number) {

    if (this.question.user_id == this.connectedUser) {
      service
        .getAnswerById(answer_id)
        .then((answer) => (this.answer = answer))
        .then(() => {
          this.answer.confirmed_answer = 1;
          service.updateAnswer(this.answer)
          .then(() => this.mounted())
          .catch((error) => console.error('Error saving answer: ' + error.message));
        
      })
    .catch((error) => console.error('Error getting answer: ' + error.message));
   }else alert("You are not the owner of this question")
  }
}
