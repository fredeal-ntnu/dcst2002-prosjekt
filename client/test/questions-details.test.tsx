// Import necessary libraries and components
import React from 'react';
import { shallow, mount } from 'enzyme';
import { QuestionDetails } from 'src/components/question-details'; // Update the path accordingly
import service from 'src/service';

// Mock the service module to simulate API calls
jest.mock('src/service', () => ({
  getMe: jest.fn(() => Promise.resolve({ user_id: 1 })),
  getAllTags: jest.fn(() => Promise.resolve([])),
  getQuestion: jest.fn(() => Promise.resolve({ question_id: 1, title: 'Test Question', text: 'Test Text' })),
  getAllTagQuestionRelations: jest.fn(() => Promise.resolve([])),
  getAnswerScoresByQuestionId: jest.fn(() => Promise.resolve([])),
  getQuestionCommentsForQuestion: jest.fn(() => Promise.resolve([])),
  // Add other mock implementations for your service methods
  // ...
}));

describe('QuestionDetails Component', () => {
  let wrapper;

  beforeEach(() => {
    // Shallow render the component before each test
    wrapper = shallow(<QuestionDetails match={{ params: { id: 1 } }} />);
  });

  it('renders without crashing', () => {
    expect(wrapper.exists()).toBe(true);
  });

  it('initializes state correctly', () => {
    expect(wrapper.state('isFavorite')).toBe(false);
    expect(wrapper.state('isConfirmedAnswer')).toBe(false);
    // Add other assertions for state properties if necessary
  });

  it('loads data on mount', async () => {
    // Mock the service methods to return data
    await wrapper.instance().mounted();

    // Expectations for data loading
    expect(service.getMe).toHaveBeenCalled();
    expect(service.getAllTags).toHaveBeenCalled();
    expect(service.getQuestion).toHaveBeenCalledWith(1);
    expect(service.getAllTagQuestionRelations).toHaveBeenCalled();
    expect(service.getAnswerScoresByQuestionId).toHaveBeenCalledWith(1);
    expect(service.getQuestionCommentsForQuestion).toHaveBeenCalledWith(1);
    // Add other expectations for data loading if necessary
  });

  // Add more tests for rendering components, handling user interactions, etc.

  it('renders question details correctly', () => {
    // Expectations for rendering question details
    expect(wrapper.find('Card[title="Title"]').text()).toBe('Test Question');
    expect(wrapper.find('Card[title="Text"]').text()).toBe('Test Text');
    // Add other expectations for rendering question details if necessary
  });

  it('handles filter change correctly', () => {
    const select = wrapper.find('Form.Select');

    // Simulate a filter change event
    select.simulate('change', { target: { value: 'popular' } });

    // Expect the component state to be updated
    expect(wrapper.state('filter')).toBe('popular');
  });

  // Add more tests for other component functionalities, such as creating comments, adding answers, etc.

  // Don't forget to test asynchronous functionalities using async/await or other mechanisms provided by testing libraries.
});
