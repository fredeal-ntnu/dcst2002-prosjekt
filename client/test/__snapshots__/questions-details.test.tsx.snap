// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site renders Site renders correct 1`] = `
<Fragment>
  <Column
    width={3}
  >
    <SideMenu
      header="Menu"
      items={
        [
          {
            "label": "Questions",
            "to": "/questions",
          },
          {
            "label": "My Questions",
            "to": "/myquestions",
          },
          {
            "label": "My Favourite Answers",
            "to": "/favourites",
          },
          {
            "label": "New Question",
            "to": "/createquestion",
          },
          {
            "label": "Tags",
            "to": "/tags",
          },
        ]
      }
    />
  </Column>
  <Card
    title="Question"
  >
    <MiniCard />
    <MiniCard
      title="Tags"
    >
      test
    </MiniCard>
    <Row>
      <Column
        width={1}
      >
        <ButtonSuccess
          onClick={[Function]}
        >
          Edit
        </ButtonSuccess>
      </Column>
    </Row>
  </Card>
  <Row>
    <Column
      width={5}
    >
      <Card
        title="Comments"
      >
        <Card
          key="1"
          title=""
        >
          <Row>
            test
          </Row>
        </Card>
      </Card>
    </Column>
    <Column
      width={5}
    >
      <Card
        title="Answers"
      >
        <FormSelect
          onChange={[Function]}
          value="all"
        >
          <option
            value="all"
          >
            All Answers
          </option>
          <option
            value="popular"
          >
            Most Popular Answers
          </option>
          <option
            value="mostRecent"
          >
            Most Recent
          </option>
          <option
            value="confirmed"
          >
            Confirmed Answers
          </option>
        </FormSelect>
      </Card>
    </Column>
  </Row>
</Fragment>
`;

exports[`switchcase switchcase confirmed 1`] = `
<Fragment>
  <Column
    width={3}
  >
    <SideMenu
      header="Menu"
      items={
        [
          {
            "label": "Questions",
            "to": "/questions",
          },
          {
            "label": "My Questions",
            "to": "/myquestions",
          },
          {
            "label": "My Favourite Answers",
            "to": "/favourites",
          },
          {
            "label": "New Question",
            "to": "/createquestion",
          },
          {
            "label": "Tags",
            "to": "/tags",
          },
        ]
      }
    />
  </Column>
  <Card
    title="Question"
  >
    <MiniCard />
    <MiniCard
      title="Tags"
    >
      test
    </MiniCard>
    <Row>
      <Column
        width={1}
      >
        <ButtonSuccess
          onClick={[Function]}
        >
          Edit
        </ButtonSuccess>
      </Column>
    </Row>
  </Card>
  <Row>
    <Column
      width={5}
    >
      <Card
        title="Comments"
      >
        <Card
          key="1"
          title=""
        >
          <Row>
            test
          </Row>
        </Card>
      </Card>
    </Column>
    <Column
      width={5}
    >
      <Card
        title="Answers"
      >
        <FormSelect
          onChange={[Function]}
          value="confirmed"
        >
          <option
            value="all"
          >
            All Answers
          </option>
          <option
            value="popular"
          >
            Most Popular Answers
          </option>
          <option
            value="mostRecent"
          >
            Most Recent
          </option>
          <option
            value="confirmed"
          >
            Confirmed Answers
          </option>
        </FormSelect>
      </Card>
    </Column>
  </Row>
</Fragment>
`;

exports[`switchcase switchcase most recent 1`] = `
<Fragment>
  <Column
    width={3}
  >
    <SideMenu
      header="Menu"
      items={
        [
          {
            "label": "Questions",
            "to": "/questions",
          },
          {
            "label": "My Questions",
            "to": "/myquestions",
          },
          {
            "label": "My Favourite Answers",
            "to": "/favourites",
          },
          {
            "label": "New Question",
            "to": "/createquestion",
          },
          {
            "label": "Tags",
            "to": "/tags",
          },
        ]
      }
    />
  </Column>
  <Card
    title="Question"
  >
    <MiniCard />
    <MiniCard
      title="Tags"
    >
      test
    </MiniCard>
    <Row>
      <Column
        width={1}
      >
        <ButtonSuccess
          onClick={[Function]}
        >
          Edit
        </ButtonSuccess>
      </Column>
    </Row>
  </Card>
  <Row>
    <Column
      width={5}
    >
      <Card
        title="Comments"
      >
        <Card
          key="1"
          title=""
        >
          <Row>
            test
          </Row>
        </Card>
      </Card>
    </Column>
    <Column
      width={5}
    >
      <Card
        title="Answers"
      >
        <FormSelect
          onChange={[Function]}
          value="mostRecent"
        >
          <option
            value="all"
          >
            All Answers
          </option>
          <option
            value="popular"
          >
            Most Popular Answers
          </option>
          <option
            value="mostRecent"
          >
            Most Recent
          </option>
          <option
            value="confirmed"
          >
            Confirmed Answers
          </option>
        </FormSelect>
      </Card>
    </Column>
  </Row>
</Fragment>
`;

exports[`switchcase switchcase popular 1`] = `
<Fragment>
  <Column
    width={3}
  >
    <SideMenu
      header="Menu"
      items={
        [
          {
            "label": "Questions",
            "to": "/questions",
          },
          {
            "label": "My Questions",
            "to": "/myquestions",
          },
          {
            "label": "My Favourite Answers",
            "to": "/favourites",
          },
          {
            "label": "New Question",
            "to": "/createquestion",
          },
          {
            "label": "Tags",
            "to": "/tags",
          },
        ]
      }
    />
  </Column>
  <Card
    title="Question"
  >
    <MiniCard />
    <MiniCard
      title="Tags"
    >
      test
    </MiniCard>
    <Row>
      <Column
        width={1}
      >
        <ButtonSuccess
          onClick={[Function]}
        >
          Edit
        </ButtonSuccess>
      </Column>
    </Row>
  </Card>
  <Row>
    <Column
      width={5}
    >
      <Card
        title="Comments"
      >
        <Card
          key="1"
          title=""
        >
          <Row>
            test
          </Row>
        </Card>
      </Card>
    </Column>
    <Column
      width={5}
    >
      <Card
        title="Answers"
      >
        <FormSelect
          onChange={[Function]}
          value="popular"
        >
          <option
            value="all"
          >
            All Answers
          </option>
          <option
            value="popular"
          >
            Most Popular Answers
          </option>
          <option
            value="mostRecent"
          >
            Most Recent
          </option>
          <option
            value="confirmed"
          >
            Confirmed Answers
          </option>
        </FormSelect>
      </Card>
    </Column>
  </Row>
</Fragment>
`;
